src/transactions/TransactionUtils.cpp

```diff
@@ -1142,34 +1142,245 @@ claimableBalanceFlagIsValid(ClaimableBalanceEntry const& cb)
+    {
+        LedgerTxn ltxInner(ltx);
+
+        // load all relevant pool share trustLines
+        auto poolShareTrustLines =
+            ltxInner.loadPoolShareTrustLinesByAccountAndAsset(accountID, asset);
+
+        if (poolShareTrustLines.empty())
+        {
+            return true;
+        }
+
+        // load all relevant pools
+        UnorderedSet<LedgerKey> poolKeys;
+        for (auto& trustLine : poolShareTrustLines)
         {
-            throw std::runtime_error(
-                "Offer not buying or selling expected asset");
+            poolKeys.emplace(liquidityPoolKey(
+                trustLine.current().data.trustLine().asset.liquidityPoolID()));
         }
+        ltxInner.prefetch(poolKeys);
 
-        releaseLiabilities(ltxInner, header, offer);
-        auto trustAcc = stellar::loadAccount(ltxInner, account);
-        removeEntryWithPossibleSponsorship(ltxInner, header, offer.current(),
-                                           trustAcc);
-        offer.erase();
+        for (auto& poolShareTrustLine : poolShareTrustLines)
+        {
+            // use a lambda so we don't hold a reference to the internals of
+            // TrustLineEntry
+            auto poolTL = [&]() -> TrustLineEntry const& {
+                return poolShareTrustLine.current().data.trustLine();
+            };
+
+            auto pool =
+                loadLiquidityPool(ltxInner, poolTL().asset.liquidityPoolID());
+            auto balance = poolTL().balance;
+            auto cbSponsoringAccID =
+                poolTL().ext.v() == 1 &&
+                        getLedgerEntryExtensionV1(poolShareTrustLine.current())
+                            .sponsoringID
+                    ? *getLedgerEntryExtensionV1(poolShareTrustLine.current())
+                           .sponsoringID
+                    : accountID;
+
+            // release reserves for pool share trustline that will be deleted
+            // later
+            {
+                auto trustAcc = stellar::loadAccount(ltxInner, accountID);
+                removeEntryWithPossibleSponsorship(
+                    ltxInner, ltxInner.loadHeader(),
+                    poolShareTrustLine.current(), trustAcc);
+            }
+
+            auto redeemIntoClaimableBalance =
+                [&ltxInner, &txSourceID, &txSeqNum, &opIndex, &poolTL,
+                 &accountID, &cbSponsoringAccID](Asset const& assetInPool,
+                                                 int64_t amount) -> bool {
+                // create a claimable balance for the asset in the pool
+                LedgerEntry claimableBalance;
+                claimableBalance.data.type(CLAIMABLE_BALANCE);
+
+                auto& claimableBalanceEntry =
+                    claimableBalance.data.claimableBalance();
+
+                OperationID operationID;
+                operationID.type(ENVELOPE_TYPE_POOL_REVOKE_OP_ID);
+                operationID.revokeId().sourceAccount = txSourceID;
```

Comment:
This doesn't match our normal naming convention. I think we should fix the CAP so that it's `revokeID` not `revokeId` (see `AccountID`, `OperationID`, `BalanceID`, etc for reference).

Reply:
Fixed